### ISSUE TITLE
insserv: Service zvol has to be enabled to start service shareiscsi

### ISSUE THREAD

**Issue by aderouineau** [zfsonlinux/zfs#2680]
I am getting the following error message when trying to install debian-zfs on Wheezy:

Setting up zfsutils (0.6.3-1~wheezy) ...
insserv: Service zvol has to be enabled to start service shareiscsi
insserv: exiting now!
update-rc.d: error: insserv rejected the script header
dpkg: error processing zfsutils (--configure):
 subprocess installed post-installation script returned error exit status 1
dpkg: dependency problems prevent configuration of debian-zfs:
 debian-zfs depends on zfsutils; however:
  Package zfsutils is not configured yet.

dpkg: error processing debian-zfs (--configure):
 dependency problems - leaving unconfigured
Processing triggers for initramfs-tools ...
update-initramfs: Generating /boot/initrd.img-2.6.32-openvz-amd64
Errors were encountered while processing:
 zfsutils
 debian-zfs
E: Sub-process /usr/bin/dpkg returned an error code (1)

**Comment by tobia** [comment]
I'm getting the same exact message, installing current releases (as of this date) onto Debian Wheezy.

Here are the requested commands and their output. The solution does not seem to work, it gives the same error message:

	# ls -l /etc/init.d/zfs-mount
	-rwxr-xr-x 1 root root 5870 Jun 27 15:04 /etc/init.d/zfs-mount
	
	# grep update-rc /var/lib/dpkg/info/zfsutils.*inst*
	/var/lib/dpkg/info/zfsutils.postinst:  update-rc.d zfs-mount defaults 13 07 >/dev/null || exit $?
	/var/lib/dpkg/info/zfsutils.postinst:  update-rc.d zfs-share defaults 18 00 >/dev/null || exit $?
	/var/lib/dpkg/info/zfsutils.postinst:  update-rc.d zed defaults >/dev/null
	
	# update-rc.d zfs-mount defaults 13 07
	update-rc.d: using dependency based boot sequencing
	insserv: Service zfs-mount has to be enabled to start service zvol
	insserv: exiting now!
	update-rc.d: error: insserv rejected the script header
	
	# update-rc.d zfs-share defaults 18 00
	update-rc.d: using dependency based boot sequencing
	insserv: Service zfs-mount has to be enabled to start service shareiscsi
	insserv: Service zvol has to be enabled to start service shareiscsi
	insserv: exiting now!
	update-rc.d: error: insserv rejected the script header

**Comment by FransUrbo** [comment]
> # update-rc.d zfs-mount defaults 13 07
> insserv: Service zfs-mount has to be enabled to start service zvol
I'm stumped! No idea why that's happening and I can't see anything obvious either.

From the size of the file, I see that this is from the "released" packages (not "dailies"). Those have been working for years and I've never seen any problem with them before. I haven't changed anything in that area that I know of either…

What does the following command reveal?

	ls -l $(find /etc/rc?.d -name '*zfs*' -o -name '*zed*')


Is this a NEW installation [of the packages], or an upgrade from a previous version? If so, from which one?

**Comment by tobia** [comment]
I have also installed the released packages on other Debian Wheezy servers in the past, so I don't know what's going on.

This is a new installation over a newly installed server. It has the Proxmox PVE packages and kernel installed, so it's not pure Debian Wheezy, but then again this is a kind of installation I have done several times in the past, without this bug. Maybe they have changed something recently in the PVE packages?

    deb http://download.proxmox.com/debian wheezy pve-no-subscription

There is currently no zfs symlink in `rc?.d`:

    # find /etc/rc?.d -name '*zfs*' -o -name '*zed*'
    #

**Comment by FransUrbo** [comment]
> Maybe they have changed something recently in the PVE packages?
> 
Maybe, but I find that doubtful. The problem here seems to be update-rc.d, and unless they messed with that specifically (or insserv), I can't see how their packages would clash with mine and give this problem.
> There is currently no zfs symlink in rc?.d:
Ok, I kind'a expected that, but just to be sure…

Because of the error:
> # update-rc.d zfs-mount defaults 13 07
> insserv: Service zfs-mount has to be enabled to start service zvol
and the header of the zfs-mount script:

	### BEGIN INIT INFO
	# Provides:       zvol zfs zfs-mount
	# Required-Start: $local_fs
	# Required-Stop:  
	# X-Stop-After:   umountfs
	# Default-Start:  2 3 4 5
	# Default-Stop:   0 1 6
	# Short-Description: Import and mount ZFS pools, filesystems and volumes
	# Description: Run the `zpool import`, `zfs mount -a`, `zfs umount -a` or
	#              `zpool export` command.
	### END INIT INFO

I can't understand why it complains. It PROVIDES zvol, it doesn't REQUIRE it (which the error seems to indicate)!

I'm stumped and have no ideas of hand to give you, but if you ever figure out the problem, or how to fix it (other than purging the package and reinstall it - which might be the only option if that actually works 100% of the times!). I can't even understand why THAT would work!

**Comment by dannyash** [comment]
I am having the same problem on Debian Jessie. Removing zfsutils and purging does not solve the problem. I ran your commands as above, and got this:

/$ sudo ls -l /etc/init.d/zfs-mount<
-rwxr-xr-x 1 root root 5870 Jun 27 09:08 /etc/init.d/zfs-mount
/$ sudo grep update-rc /var/lib/dpkg/info/zfsutils.*inst=
grep: /var/lib/dpkg/info/zfsutils.*inst=: No such file or directory</code>

Is that any help?

**Comment by FransUrbo** [comment]
> /$ sudo grep update-rc /var/lib/dpkg/info/zfsutils.inst=
> grep: /var/lib/dpkg/info/zfsutils.inst=: No such file or directory
> 

That command should have ended with: *inst*

As in: STAR inst STAR (without spaces). NOT: STAR inst EQUALSIGN.

@dannyash. Please run the correct commands and then give me the exact output (via cut and paste).

**Comment by wgogan** [comment]
I'm experiencing this too.. here's what I get:
```
> grep update-rc /var/lib/dpkg/info/zfsutils.*inst*
/var/lib/dpkg/info/zfsutils.postinst:	update-rc.d zfs-zed start 07 S . stop 08 0 1 6 . >/dev/null || exit $?
/var/lib/dpkg/info/zfsutils.postinst:	update-rc.d zfs-import start 07 S . stop 07 0 1 6 . >/dev/null || exit $?
/var/lib/dpkg/info/zfsutils.postinst:	update-rc.d zfs-mount start 02 2 3 4 5 . stop 06 0 1 6 . >/dev/null || exit $?
/var/lib/dpkg/info/zfsutils.postinst:	update-rc.d zfs-share start 27 2 3 4 5 . stop 05 0 1 6 . >/dev/null || exit $?
```

```
# ls -l $(find /etc/rc?.d -name '*zfs*' -o -name '*zed*')
total 1992
-rw-r--r-- 1 root root 2039658 Apr 18 17:23 zfsonlinux_6_all.deb
```

```# ls -l /etc/init.d/zfs-mount
-rwxr-xr-x 1 root root 5362 Sep 17 19:26 /etc/init.d/zfs-mount
```

Hope this helps in tracking it down. I've tried various remove/purge/reinstall combinations, and nothing has managed to get this working for me yet on Debian.

**Comment by Blackclaws** [comment]
I have the same problem I set up a brand new debian jessie on my new server and tried to install debian-zfs it didn't work with similar errors. I'm going to post my errors to see if that can help anyone track the error down:

DKMS: install completed.
Setting up menu (2.1.47) ...
Setting up spl (0.6.5-1) ...
Setting up zfsutils (0.6.5-3) ...
update-rc.d: warning: start and stop actions are no longer supported; falling back to defaults
update-rc.d: warning: start runlevel arguments (S) do not match zfs-zed Default-Start values (2 3 4 5)
insserv: Service zfs-mount has to be enabled to start service zfs-zed
insserv: exiting now!
update-rc.d: error: insserv rejected the script header
dpkg: error processing package zfsutils (--configure):
 subprocess installed post-installation script returned error exit status 1
Processing triggers for menu (2.1.47) ...
Processing triggers for initramfs-tools (0.120) ...
update-initramfs: Generating /boot/initrd.img-3.16.0-4-amd64
Processing triggers for systemd (215-17+deb8u2) ...
Errors were encountered while processing:
 zfsutils
E: Sub-process /usr/bin/dpkg returned an error code (1)

Then the first set of commands

root@homeserver:/home/felix# ls -l /etc/init.d/zfs-mount
-rwxr-xr-x 1 root root 5362 Sep 18 17:10 /etc/init.d/zfs-mount
root@homeserver:/home/felix# grep update-rc /var/lib/dpkg/info/zfsutils.*inst*
/var/lib/dpkg/info/zfsutils.postinst:   update-rc.d zfs-zed start 07 S . stop 08 0 1 6 . >/dev/null || exit $?
/var/lib/dpkg/info/zfsutils.postinst:   update-rc.d zfs-import start 07 S . stop 07 0 1 6 . >/dev/null || exit $?
/var/lib/dpkg/info/zfsutils.postinst:   update-rc.d zfs-mount start 02 2 3 4 5 . stop 06 0 1 6 . >/dev/null || exit $?
/var/lib/dpkg/info/zfsutils.postinst:   update-rc.d zfs-share start 27 2 3 4 5 . stop 05 0 1 6 . >/dev/null || exit $?


Also:

root@homeserver:/home/felix# update-rc.d zfs-mount defaults 13 07 && update-rc.d zfs-share defaults 18 00
insserv: Service add has to be enabled to start service zfs-mount
insserv: Service zfs-mount has to be enabled to start service zfs-mount
insserv: Service zfs-import has to be enabled to start service zfs-mount
insserv: exiting now!
update-rc.d: error: insserv rejected the script header


I have absolutely no idea whats going on here

**Comment by subzero79** [comment]
The zfs-zed has in its postinst runlevel S, however the default runlevels of the init lsb tags are 2345,

Another thing i don't know how the zfsutils are packaged but also the order of the postinst seems wrong.

should be zfs-import, zfs-mount, zfs-zed, zfs-share, it has currently zfs-zed in first place when it depends of zfs-mount

**Comment by TonyHoyle** [comment]
I got a working system by using:

update-rc.d zfs-import start 07 S . stop 07 0 1 6 .
update-rc.d zfs-mount start 02 2 3 4 5 . stop 06 0 1 6 .
update-rc.d zfs-zed start 07 2 3 4 5 . stop 08 0 1 6 .
update-rc.d zfs-share start 27 2 3 4 5 . stop 05 0 1 6 .

**Comment by kcook769** [comment]
On jessie I edited zfsutils.postinst and moved zfs-zed  to occur below zfs-mount and zfs-import which fixed the issue for me.

I believe as long as zfs-zed's update-rc.d section is listed after zfs-mount in postinst that should resolve the issue since zfs-zed depends on zfs-mount (LSB headers)

### BEGIN INIT INFO
# Provides:          zfs-zed
# Required-Start:    zfs-mount
# Required-Stop:     zfs-mount

**Comment by FransUrbo** [comment]
I'm fairly certain that this is a pkg-zfs issue, NOT zfs. I missed to have it moved when it was new.

And I think I know the cause to, I just haven't had time to deal with it.

**Comment by FransUrbo** [comment]
Looking closer, it's the 3f1cc17c9005d19faba91fe7236332ed502b1a3f commit - the commands it changed in the <code>README.md</code> is wrong.

I'm including a fix for this in my cherry-pick PR #3816.

**Comment by FransUrbo** [comment]
This last problem is (should) now fixed in the <code>0.6.5.1-3</code> version.

As to the original problem, the iscsi problem, this might still be a problem. @behlendorf I suggest closing this (due to the new SYSV init scripts) AND the fact that it's been … "tainted" (with other, possibly non related things). If this (the original issue) should still a problem, someone can create a new issue and we can take it from there.

**Comment by FransUrbo** [comment]
Second part of the fix is in <code>0.6.5.1-4</code>. Please report success/failure as soon as possible so we can close this.

**Comment by tobia** [comment]
For what concerns my report above, I had made a mistake in the init sequence of ZFS, so it might be unrelated to the other errors reported here. I managed to fix it this way: 
* in `/etc/insserv.conf` I appended `zfs-mount` (without a `+`) to `$local_fs`
* in `/etc/init.d/zfs-mount` I removed `Required-Start: $local_fs` and set `Default-Start: S`
* then I rebuilt the sequence with `insserv -v -d zfs-mount`

**Comment by FransUrbo** [comment]
<code>0.6.5.1-4</code> is verified to work and install fresh without problems.

**Comment by behlendorf** [comment]
Closing, addressed by 0c8a4cf
